data_structures/linked_list/singly_link_list_deletion.c

```diff
@@ -2,97 +2,136 @@
   It is assumed that the data in nodes will be an integer, though
   function can be modified according to the data type, easily.
   deleteNode deletes a node when passed with a key of the node.
+  The linked list can be used as a stack by implementing the functions insertLeft and deleteLeft
+  The linked list can be used as a queue by implementing the functions insertRight and deleteLeft
+  The linked list can be used as a double ended queue(dequeue) by implementing the functions insertRight and insertLeft to insert nodes on both sides and deleteLeft and deleteRight to delete nodes on both sides
 */
+///////////////////////////////////////////////////////////
 #include<stdio.h>
-struct node
-{int info;
+#include<stdlib.h>
+typedef struct node
+{
+ int info;
  struct node *link;
-};
-struct node *start=NULL;
+}*NODE;
+///////////////////////////////////////////////////////////
 ///////////////////////////////////////////////////////////
-struct node * createnode()//function to create node
+NODE insertLeft(NODE start) //function to insert the new node towards the right and function to create node as well
 {
-  struct node *t;
-  t=(struct node*)malloc(sizeof(struct node));
-  return(t);
+  NODE p;
+  p=(NODE)malloc(sizeof(struct node));
+  printf("enter the number to insert\n");
+  scanf("%d",&(p->info));
+  p->link=NULL;
+  if(start==NULL)
+  {
+      start=p;
+      return start;
+  }
+  p->link=start;
+  start=p;
+  return start;
+
 }
-////////////////////////////////////////////////////////
-void insert()//function to insert at first location
+///////////////////////////////////////////////////////////
+NODE insertRight(NODE start)
 {
-  struct node *p;
-  p=createnode();
-  printf("\nenter the number to insert");
-  scanf("%d",&p->info);
+  NODE p;
+  p=(NODE)malloc(sizeof(struct node));
+  printf("enter the number to be inserted\n");
+  scanf("%d",&(p->info));
   p->link=NULL;
   if(start==NULL)
   {
-      start=p;
+    start=p;
+    return start;
   }
-  else
+  NODE temp=start;
+  while(temp->link!=NULL)
   {
-      p->link=start;
-      start=p;
+    temp=temp->link;
   }
+  temp->link=p;
+  return start;
+
 }
 ///////////////////////////////////////////////////////////
-void deleteion()//function to delete from first position
+NODE deleteLeft(NODE start)//function to delete from the left
 {
-    struct node *t;
+    NODE t=NULL;
     if(start==NULL)
     {
-        printf("\nlist is empty");
-    }
-    else
-    {
-        struct node *p;
-        p=start;
-        start=start->link;
-        free(p);
+        printf("list is empty\n");
+        return start;
     }
+    t=start;
+    start=start->link;
+    printf("number deleted is: %d",t->info);
+    free(t);
+    return start;
+}
+///////////////////////////////////////////////////////
+NODE deleteRight(NODE start)
+{
+  NODE t=start,prev=NULL;
+  while(t->link!=NULL)
+  {
+    prev=t;
+    t=t->link;
+  }
+  printf("number deleted is: %d",t->info);
+  free(t);
+  prev->link=NULL;
```

Comment:
if there is only 1 node in the linked list, ``prev`` will be ``NULL`` and ``prev->link`` will cause error